### PR TITLE
includes origin's stake in gossip push nodes sampling

### PR DESCRIPTION
**Pull request by behzadnouri**
#### Problem
Gossip push samples nodes by stake. This is unnecessarily wasteful and creates too much congestion at high staked nodes if the CRDS value to be propagated is from a node with low or zero stake.

#### Summary of Changes
This commit instead maintains several active-sets for push, each corresponding with a stake bucket. Peer sampling weights are accordingly capped by the respective bucket stake.

### CODE REVIEW

File: gossip/src/push_active_set.rs [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+        for origin in origins {
+            if origin == pubkey {
+                continue;
+            }
+            let stake = stake.min(stakes.get(origin));
+            self.get_entry(stake).prune(node, origin)
+        }
+    }
+
+    pub(crate) fn rotate<R: Rng>(
```
Comment (t-nelson) [inline gossip/src/push_active_set.rs:54]: "rotate" isn't an intuitive operation here in my mind. can't decide whether a rename or doc comment is the solution though :thinking:

File: gossip/src/push_active_set.rs [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+        entry.rotate(&mut rng, 6, NUM_BLOOM_FILTER_ITEMS, &nodes, &weights);
+        let keys = [
+            &nodes[17], &nodes[14], &nodes[5], &nodes[7], &nodes[1], &nodes[13],
+        ];
+        assert!(entry.0.keys().eq(keys));
+        entry.rotate(&mut rng, 4, NUM_BLOOM_FILTER_ITEMS, &nodes, &weights);
+        let keys = [&nodes[5], &nodes[7], &nodes[1], &nodes[13]];
+        assert!(entry.0.keys().eq(keys));
+    }
+}
```
Comment (t-nelson) [inline gossip/src/push_active_set.rs:237]: `PushActiveSet` could use some coverage

File: gossip/src/push_active_set.rs [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+            .map(|node| {
+                let stake = stakes.get(node).copied().unwrap_or_default();
+                let stake = stake / LAMPORTS_PER_SOL;
+                (u64::BITS - stake.leading_zeros()) as u64
+            })
+            .collect();
+        for (k, entry) in self.0.iter_mut().enumerate() {
+            let weights: Vec<u64> = weights
+                .iter()
+                .map(|&weight| weight.min(k as u64).saturating_add(1).saturating_pow(2))
```
Comment (t-nelson) [inline gossip/src/push_active_set.rs:74]: not sure i understand the math here.  a comment might help

File: gossip/src/push_active_set.rs [hunk 4]
```diff
@@ -0,0 +1,237 @@
+use {
+    crate::weighted_shuffle::WeightedShuffle,
+    indexmap::IndexMap,
+    rand::Rng,
+    solana_bloom::bloom::{AtomicBloom, Bloom},
+    solana_sdk::{native_token::LAMPORTS_PER_SOL, pubkey::Pubkey},
+    std::collections::HashMap,
+};
+
+const NUM_PUSH_ACTIVE_SET_ENTRIES: usize = 36;
```
[the remark this answers is not in the document]
Comment (behzadnouri) [reply gossip/src/push_active_set.rs:12]: It can, but it won't be `const`.
Added some comments where this value comes from.
Also this is test in `test_get_stake_bucket`.

File: gossip/src/push_active_set.rs [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+#[derive(Default)]
+struct PushActiveSetEntry(IndexMap</*node:*/ Pubkey, /*origins:*/ AtomicBloom<Pubkey>>);
+
+impl PushActiveSet {
+    #[cfg(debug_assertions)]
+    const MIN_NUM_BLOOM_ITEMS: usize = 512;
+    #[cfg(not(debug_assertions))]
+    const MIN_NUM_BLOOM_ITEMS: usize = crate::cluster_info::CRDS_UNIQUE_PUBKEY_CAPACITY;
+
+    pub(crate) fn get<'a>(
```
[the remark this answers is not in the document]
Comment (behzadnouri) [reply gossip/src/push_active_set.rs]: Renamed to `get_nodes`.

File: gossip/src/push_active_set.rs [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+        nodes: &[Pubkey],
+        stakes: &HashMap<Pubkey, u64>,
+    ) {
+        let num_bloom_filter_items = cluster_size.max(Self::MIN_NUM_BLOOM_ITEMS);
+        let weights: Vec<_> = nodes
+            .iter()
+            .map(|node| {
+                let stake = stakes.get(node).copied().unwrap_or_default();
+                let stake = stake / LAMPORTS_PER_SOL;
+                (u64::BITS - stake.leading_zeros()) as u64
```
[the remark this answers is not in the document]
Comment (behzadnouri) [reply gossip/src/push_active_set.rs]: done: `get_stake_bucket`.

File: gossip/src/push_active_set.rs [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+
+    pub(crate) fn get<'a>(
+        &'a self,
+        pubkey: &Pubkey,    // This node.
+        origin: &'a Pubkey, // CRDS value owner.
+        // If true forces gossip push even if the node has pruned the origin.
+        should_force_push: impl FnMut(&Pubkey) -> bool + 'a,
+        stakes: &HashMap<Pubkey, u64>,
+    ) -> impl Iterator<Item = &Pubkey> + 'a {
+        let stake = stakes.get(pubkey).min(stakes.get(origin));
```
[the remark this answers is not in the document]
Comment (behzadnouri) [reply gossip/src/push_active_set.rs:35]: Added a helper for the other one; but this fine seems fine to me to keep inline.

File: gossip/src/push_active_set.rs (hunk 1 above)
Comment (behzadnouri) [reply gossip/src/push_active_set.rs:56]: Rotate as in "key rotation".
It discards the oldest entry and adds a new one.
It also rotates entries: `[a, b, c, d]` becomes `[b, c, d, e]`.

File: gossip/src/push_active_set.rs (hunk 3 above)
Comment (behzadnouri) [reply gossip/src/push_active_set.rs]: Added comments:
```
bucket <- get_stake_bucket(min stake of {
   this node, crds value owner and gossip peer
})
weight <- (bucket + 1)^2
```
`min stake of {...}` is a proxy for how much we care about this link, and tries to mirror similar logic on the receiving end when pruning incoming links:
https://github.com/solana-labs/solana/blob/81394cf92/gossip/src/received_cache.rs#L100-L105

It is pretty heuristic but we need something here and we can re-evaluate and adjust.

File: gossip/src/push_active_set.rs [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+                .iter()
+                .map(|&weight| weight.min(k as u64).saturating_add(1).saturating_pow(2))
+                .collect();
+            entry.rotate(rng, size, num_bloom_filter_items, nodes, &weights);
+        }
+    }
+
+    fn get_entry(&self, stake: Option<&u64>) -> &PushActiveSetEntry {
+        let stake = stake.copied().unwrap_or_default() / LAMPORTS_PER_SOL;
+        let index = u64::BITS - stake.leading_zeros();
```
[the remark this answers is not in the document]
Comment (behzadnouri) [reply gossip/src/push_active_set.rs]: uhhh doesn't seem to me it matters much either way.

File: gossip/src/push_active_set.rs [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+        size: usize, // Number of nodes to retain.
+        num_bloom_filter_items: usize,
+        nodes: &[Pubkey],
+        weights: &[u64],
+    ) {
+        debug_assert_eq!(nodes.len(), weights.len());
+        debug_assert_eq!(weights.iter().find(|&&weight| weight == 0), None);
+        let shuffle = WeightedShuffle::new("rotate-active-set", weights).shuffle(rng);
+        for node in shuffle.map(|k| &nodes[k]) {
+            if self.0.len() > size {
```
[the remark this answers is not in the document]
Comment (behzadnouri) [reply gossip/src/push_active_set.rs:143]: It actually needs to be `> size`.
It is discarding the oldest entry and adding a new one.
`[a, b, c, d]` becomes `[b, c, d, e]`.

File: gossip/src/push_active_set.rs (hunk 2 above)
Comment (behzadnouri) [reply gossip/src/push_active_set.rs:310]: done.

File: gossip/src/push_active_set.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (t-nelson) [reply gossip/src/push_active_set.rs:143]: ah! maybe a comment that we intend to discard the top entry once full?

File: gossip/src/push_active_set.rs (hunk 3 above)
Comment (t-nelson) [reply gossip/src/push_active_set.rs]: thanks that helps!  mind adding this part to the code comment?

> a proxy for how much we care about this link, and tries to mirror similar logic on the receiving end when pruning incoming

File: gossip/src/push_active_set.rs (hunk 1 above)
Comment (behzadnouri) [reply gossip/src/push_active_set.rs:56]: We need `IndexMap` because we want to preserve a consistent order for the nodes which are pushed to. Similar to current master code:
https://github.com/solana-labs/solana/blob/1cc8917b9/gossip/src/crds_gossip_push.rs#L219-L224

So if `fanout = 2` and the entries are `[a, b, c, d]`, we want to push to `a` and `b`.
But if `b` gets pruned we want to push to `a` and `c`.
Then when this gets rotated to `[b, c, d, e]`, we want to push to `c` and `d`.

File: gossip/src/push_active_set.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (behzadnouri) [reply gossip/src/push_active_set.rs:148]: added comment.

File: gossip/src/push_active_set.rs (hunk 3 above)
Comment (behzadnouri) [reply gossip/src/push_active_set.rs]: done.

File: gossip/src/push_active_set.rs [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+                .map(|&bucket| {
+                    // bucket <- get_stake_bucket(min stake of {
+                    //  this node, crds value owner and gossip peer
+                    // })
+                    // weight <- (bucket + 1)^2
+                    // min stake of {...} is a proxy for how much we care about
+                    // the link, and tries to mirror similar logic on the
+                    // receiving end when pruning incoming links:
+                    // https://github.com/solana-labs/solana/blob/81394cf92/gossip/src/received_cache.rs#L100-L105
+                    let bucket = bucket.min(k) as u64;
```
Comment (carllin) [inline gossip/src/push_active_set.rs:81]: hmmm I'm still not sure I understand the significance of this min or `k`. Is there something significant about the order in which we're iterating the buckets in the outer loop `self.0.iter_mut().enumerate() {`?

File: gossip/src/push_active_set.rs [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+
+    pub(crate) fn get_nodes<'a>(
+        &'a self,
+        pubkey: &Pubkey,    // This node.
+        origin: &'a Pubkey, // CRDS value owner.
+        // If true forces gossip push even if the node has pruned the origin.
+        should_force_push: impl FnMut(&Pubkey) -> bool + 'a,
+        stakes: &HashMap<Pubkey, u64>,
+    ) -> impl Iterator<Item = &Pubkey> + 'a {
+        let stake = stakes.get(pubkey).min(stakes.get(origin));
```
Comment (carllin) [inline gossip/src/push_active_set.rs:35]: what's the reasoning for the min here instead of just using the `pubkey`'s stake? 

Is the goal to deprioritize messages from lowly staked `origin`'s as well?

File: gossip/src/push_active_set.rs [hunk 12]
```diff
@@ -0,0 +1,316 @@
...
+                    // weight <- (bucket + 1)^2
+                    // min stake of {...} is a proxy for how much we care about
+                    // the link, and tries to mirror similar logic on the
+                    // receiving end when pruning incoming links:
+                    // https://github.com/solana-labs/solana/blob/81394cf92/gossip/src/received_cache.rs#L100-L105
+                    let bucket = bucket.min(k) as u64;
+                    bucket.saturating_add(1).saturating_pow(2)
+                })
+                .collect();
+            entry.rotate(rng, size, num_bloom_filter_items, nodes, &weights);
```
Comment (carllin) [inline gossip/src/push_active_set.rs:85]: Question since the old rotation code is quite hard to read. How does this new rotation logic compare/is it significantly different?

File: gossip/src/push_active_set.rs (hunk 11 above)
Comment (behzadnouri) [reply gossip/src/push_active_set.rs:35]: On the receiving end of gossip push messages:
https://github.com/solana-labs/solana/blob/v1.14.11/gossip/src/crds_gossip_push.rs#L192
we keep enough incoming links to hit a `prune_stake_threshold` but that threshold is a function of the origin's stake:
https://github.com/solana-labs/solana/blob/v1.14.11/gossip/src/crds_gossip_push.rs#L117-L120
So if the origin is unstaked or has low stake, we don't really care much about the stake of the incoming link.

So here on the push side, if the value is from an untaked or low staked `origin` then, instead of sampling nodes by stake, we may utilize the wider cluster to propagate it more efficiently.

> Is the goal to deprioritize messages from lowly staked origin's as well?

Not to deprioritize.
But instead of tunneling them through the high staked nodes (which would unnecessarily overload and create congestion at high staked nodes), use the wider cluster to propagate those.
For these values, the receiving node does not care if the incoming link has high stake or not, so we may sample the cluster more uniformly.

File: gossip/src/push_active_set.rs (hunk 12 above)
Comment (behzadnouri) [reply gossip/src/push_active_set.rs:85]: Instead of these weights:
https://github.com/solana-labs/solana/blob/v1.14.11/gossip/src/crds_gossip_push.rs#L465-L471
I am using weights as calculated above.

When discarding entries:
https://github.com/solana-labs/solana/blob/v1.14.11/gossip/src/crds_gossip_push.rs#L418-L420
I am preserving the order of nodes in `IndexMap`.

Otherwise should be functionally the same.

File: gossip/src/push_active_set.rs (hunk 10 above)
Comment (behzadnouri) [reply gossip/src/push_active_set.rs:81]: Added some more context in the other comment:

> On the receiving end of gossip push messages:
> https://github.com/solana-labs/solana/blob/v1.14.11/gossip/src/crds_gossip_push.rs#L192
> we keep enough incoming links to hit a `prune_stake_threshold` but that threshold is a function of the origin's stake:
> https://github.com/solana-labs/solana/blob/v1.14.11/gossip/src/crds_gossip_push.rs#L117-L120
> So if the origin is unstaked or has low stake, we don't really care much about the stake of the incoming link.

We want
```
get_stake_bucket(min stake of {
   this node, crds value owner and gossip peer
})
```
as a proxy for how much we care about this link.
Here, stakes are mapped to 0..n buckets, and each bucket has its own active-set.
The k'th bucket is for origins which `min stake of {this node and the crds value owner}` maps to bucket `k`, i.e. the first 2 items in above. Taking the `min` with node's bucket results in the whole thing.

File: gossip/src/push_active_set.rs (hunk 12 above)
Comment (carllin) [reply gossip/src/push_active_set.rs:85]: got it, yeah that old code based on when the node was last pushed to seems kind of hacky 😝

File: gossip/src/push_active_set.rs (hunk 10 above)
Comment (behzadnouri) [reply gossip/src/push_active_set.rs:93]: The suggested comments were actually the other way around.
Each `PushActiveSetEntry` corresponds to a stake bucket for
```
min stake of { this node, crds value owner }
```
The entry represents set of gossip nodes to actively push to for crds values belonging to the bucket.
`nodes: &[Pubkey]` are gossip nodes to be sampled for each push active set.
I have added further comments to the code as below.
```rust
// Active set of nodes to push to are sampled from these gossip nodes,
// using sampling probabilities obtained from the stake bucket of each
// node.
let buckets: Vec<_> = nodes
    .iter()
    .map(|node| get_stake_bucket(stakes.get(node)))
    .collect();
// (k, entry) represents push active set where the stake bucket of
//     min stake of {this node, crds value owner}
// is equal to `k`. The `entry` maintains set of gossip nodes to
// actively push to for crds values belonging to this bucket.
for (k, entry) in self.0.iter_mut().enumerate() {
    let weights: Vec<u64> = buckets
        .iter()
        .map(|&bucket| {
            // bucket <- get_stake_bucket(min stake of {
            //  this node, crds value owner and gossip peer
            // })
            // weight <- (bucket + 1)^2
            // min stake of {...} is a proxy for how much we care about
            // the link, and tries to mirror similar logic on the
            // receiving end when pruning incoming links:
            // https://github.com/solana-labs/solana/blob/81394cf92/gossip/src/received_cache.rs#L100-L105
            let bucket = bucket.min(k) as u64;
            bucket.saturating_add(1).saturating_pow(2)
        })
        .collect();
    entry.rotate(rng, size, num_bloom_filter_items, nodes, &weights);
}
```

File: gossip/src/push_active_set.rs (hunk 11 above)
Comment (behzadnouri) [reply gossip/src/push_active_set.rs:37]: > so the idea is to use nodes in the same bucket to propagate those messages from low staked origin
> but the messages will probably propagate through other lowly staked nodes in the same bucket?

Effectively
* For low staked origins we sample gossip nodes to push to uniformly, regardless of my node's stake.
* Similarly, if my node has low stake, then we sample gossip nodes to push to almost uniformly regardless of the origin's stake.
* But if my node has high stake and I receive a crds value from a high staked origin, then I would tend to push that value to another high staked node by giving those nodes higher sampling weight.

In other words, the only time the stakes matter is when both my node and the crds value origin have high stake. Otherwise gossip nodes to push to are sampled almost uniformly.

Again this is trying to mirror the similar logic on the receiving end:
https://github.com/solana-labs/solana/blob/v1.14.11/gossip/src/crds_gossip_push.rs#L192
https://github.com/solana-labs/solana/blob/v1.14.11/gossip/src/crds_gossip_push.rs#L117-L120
where if my node has high stake, for crds values from a high staked origin, I would have a high `prune_stake_threshold` and therefore would prefer to receive these values from a high staked node.

File: gossip/src/push_active_set.rs (hunk 10 above)
Comment (carllin) [reply gossip/src/push_active_set.rs:93]: > The entry represents set of gossip nodes to actively push to for crds values belonging to the bucket.
nodes: &[Pubkey] are gossip nodes to be sampled for each push active set.
I have added further comments to the code as below.

Ah thanks, that makes sense

**Review by carllin — DISMISSED** [review]
